File: javascript/80-implement_a_hash_table.js
class HashTable {
    constructor(size) {
        this.data = new Array(size);
    }

    _hash(key) {
        let hash = 0;
        for (let i=0; i<key.length; i++) {
            hash = (hash + key.charCodeAt(i) * i) % this.data.length;
        }
        return hash;
    }

    set(key, value) {
        let key_index = this._hash(key);
        if (!this.data[key_index]) {
            this.data[key_index] = [];
        }
        let key_already_exists = false;
        for (let i=0; i<this.data[key_index].length; i++) {
            if (this.data[key_index][i][0] === key) {
                this.data[key_index][i][1] = value;
                key_already_exists = true;
                break
            }
        }
        if (!key_already_exists) {
            this.data[key_index].push([key, value])
        }
    }

    get(key) {
        let key_index = this._hash(key);
        if (this.data[key_index]) {
            for (let i = 0; i < this.data[key_index].length; i++) {
                if (this.data[key_index][i][0] === key) {
                    return this.data[key_index][i][1]
                }
            }
        }
    }

    keys() {
        let keys_list = []
        for (let i=0; i<this.data.length; i++) {
            if (this.data[i]) {
                for (let j=0; j<this.data[i].length; j++)
                    keys_list.push(this.data[i][j][0])

            }
        }
        return keys_list
    }
}

const myHashTable = new HashTable(50);

myHashTable.set('grapes', 10000);
console.log(myHashTable.get('grapes'));

myHashTable.set('flowers', 40);
myHashTable.set('bread', 200);
console.log(myHashTable.keys());

File: python/80-implement_a_hash_table.py
class HashTable:

    def __init__(self, size):
        self.data = [None for _ in range(size)]

    @staticmethod
    def __hash(key):
        hash_value = 0
        for i, character in enumerate(str(key)):
            hash_value = (hash_value + (ord(character) * i)) % len(key)
        return hash_value

    def set(self, key, value):
        key_index = self.__hash(key)
        if not self.data[key_index]:
            self.data[key_index] = list()
            self.data[key_index].append([key, value])
        else:
            # check for hash collision
            key_already_exists = False
            for i, key_value_pair in enumerate(self.data[key_index]):
                if key_value_pair[0] == key:
                    self.data[key_index][i][1] = value
                    key_already_exists = True
                    break
            # store name if collision
            if not key_already_exists:
                self.data[key_index].append([key, value])

    def get(self, key):
        key_index = self.__hash(key)
        if self.data[key_index]:
            for key_value_pair in self.data[key_index]:
                if key == key_value_pair[0]:
                    return key_value_pair[1]

    def keys(self):
        keys_list = []
        for hash_location in self.data:
            if hash_location:
                for key_value_pair in hash_location:
                    keys_list.append(key_value_pair[0])
        return keys_list


myHashTable = HashTable(50)

myHashTable.set('grapes', 10000)
print(myHashTable.get('grapes'))

myHashTable.set('bread', 100)
myHashTable.set('wine', 50)
print(myHashTable.keys())
